db/warehouse/migrate/20230406183420_client_mci_id.rb

```diff
@@ -0,0 +1,36 @@
+class ClientMciId < ActiveRecord::Migration[6.1]
+  def change
+    create_table :external_request_logs do |t|
+      t.references :initiator, polymorphic: true
+
+      t.string :identifier
+      # t.jsonb :credentials #(user, token)
+
+      t.string :content_type
+      t.string :url, null: false
+      t.string :http_method, null: false, default: 'GET'
+      t.inet :ip
+      t.jsonb :request_headers, default: {}, null: false
+      t.text :request, null: false
+      t.text :response, null: false
+      t.integer :http_status
+      t.datetime :requested_at, null: false
+      t.timestamps
+
+      t.index :requested_at
+      t.index :ip
+      t.index [:initiator_id, :initiator_type]
+    end
+
+    create_table :external_ids do |t|
+      t.string :value, null: false
+      t.references :source, polymorphic: true, null: false # (Client, Project, Organization, etc.)
+      t.references :remote_credential, foreign_key: true  # (MCI, MPER, etc.)
+      t.references :external_request_log, foreign_key: true
+      t.timestamps
+
+      t.index :value
+      t.index [:source_id, :source_type, :remote_credential_id, :value], unique: true, name: "external_ids_uniq"
```

Comment:
@yakloinsteak 
If I understand correctly, it looks like this would allow duplicate external_ids if they reference different sources. For example, two client records (sources) could have external_id records that have the same remote_credential/ value.

I imagine we shouldn't allow that? Maybe we could have one constraint for remote_credential and one for source, something like this?
```ruby
     # a source can only have one external_id for a given remote_credential
      t.index [:source_id, :source_type, :remote_credential_id], unique: true, name: "external_ids_uniq_1"
     # remote_credential ids must be unique for a given source_type
      t.index [:source_type, :remote_credential_id, :value], unique: true, name: "external_ids_uniq_2"
```

Reply:
Feel free to make those changes if @gigxz agrees. Maybe a remote system could have multiple IDs for a source? I don't know the landscape well enough, but I think what you're saying makes sense. Definitely :+1: for more constraints when you can.

Reply:
I believe the MCI might have more than one ID per person.  That was unclear to me.  I believe **remote_credential ids must be unique for a given source_type** is a hard-fast rule.

Reply:
No, that's not quite right. At least for MCI:
1. One Client in HMIS has at most 1 MCI ID
2. Two Clients in HMIS can share the same MCI ID

Given (2), we do not want the constraint "remote_credential ids must be unique for a given source_type".

Reply:
Hmm, ok, I thought there were situations where someone was merged but the two clients had been previously given different MCI IDs so they may have more than one...

Reply:
They confirmed the above at Mondays meeting. But maybe we should ask again in the context of merging clients specifically. We haven't had an in depth conversation around merging yet.

Reply:
> Two Clients in HMIS can share the same MCI ID

@gigxz I see- so if we receive a referral request for a household member and need to match to an existing client using the MCI ID, we could find more than one matching record. I guess in that case we could try and match on secondary fields (name, social, dob). And then we'd have to use ID order as a fallback.

Reply:
Don't worry about it for now, you can just take the first matching client.